wallet/datastore.go

```diff
@@ -431,47 +357,710 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
+	}
+
+	if !noContextTx {
+		// commit
+		err = tx.Commit()
+		if err != nil {
+			sublogger.Error().Err(err).Msg("failed to commit transaction")
+			return fmt.Errorf("failed to commit transaction: %w", err)
+		}
 	}
 	return nil
 }
+
+// helper to make logger easier
+func logger(ctx context.Context) *zerolog.Logger {
+	// get logger
+	logger, err := appctx.GetLogger(ctx)
+	if err != nil {
+		// no logger, setup
+		_, logger = logging.SetupLogger(ctx)
+	}
+	return logger
+}
+
+// ConnectCustodialWallet - connect the wallet's custodial verified wallet.
+func (pg *Postgres) ConnectCustodialWallet(ctx context.Context, cl CustodianLink) error {
+	var err error
+	// create a sublogger
+	sublogger := logger(ctx).With().
+		Str("wallet_id", cl.GetWalletIDString()).
+		Str("custodian", cl.Custodian).
+		Str("linking_id", cl.GetLinkingIDString()).
+		Logger()
+
+	sublogger.Debug().
+		Msg("creating linking of wallet custodian")
+
+	// create tx
+	// get tx
+	tx, noContextTx := ctx.Value(appctx.DatabaseTransactionCTXKey).(*sqlx.Tx)
+	if !noContextTx {
+		tx, err = createTx(ctx, pg)
+		if err != nil || tx == nil {
+			sublogger.Error().Err(err).
+				Msg("error creating tx for wallet linking")
+			return fmt.Errorf("failed to create tx for wallet linking: %w", err)
+		}
+		// add tx to ctx for future
+		ctx = context.WithValue(ctx, appctx.DatabaseTransactionCTXKey, tx)
+		defer pg.RollbackTx(tx)
+	}
+
+	// if cl is not defined, this must be a new wallet linking attempt
+	if cl.ID == nil {
+		// insert wallet custodian (which will update with an ID)
+		err := pg.InsertCustodianLink(ctx, &cl)
+		if err != nil {
+			sublogger.Error().Err(err).
+				Msg("error inserting wallet linking")
+			return fmt.Errorf("failed to insert the wallet linking: %w", err)
+		}
+	}
+	sublogger.Debug().Msg("inserted custodian link")
+	// link the new wallet custodian to the wallet record in db
+	stmt := `
+		update wallets set wallet_custodian_id=$2 where id=$1
+	`
+	// perform query
+	if r, err := tx.ExecContext(
+		ctx,
+		stmt,
+		cl.WalletID,
+		cl.ID,
+	); err != nil {
+		sublogger.Error().Err(err).
+			Str("wallet_custodian_id", cl.ID.String()).
+			Msg("failed to update wallet_custodian_id for wallet")
+		return fmt.Errorf("error updating wallet with new wallet_custodian: %w", err)
+	} else if r != nil {
+		count, _ := r.RowsAffected()
+		if count < 1 {
+			sublogger.Error().Msg("at least one record should be updated for disconnecting a verified wallet")
+			return errors.New("should have updated at least one wallet with disconnected custodial")
+		}
+	}
+
+	if !noContextTx {
+		// commit
+		err = tx.Commit()
+		if err != nil {
+			sublogger.Error().Err(err).Msg("failed to commit transaction")
+			return fmt.Errorf("failed to commit transaction: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// helper to create a tx
+func createTx(ctx context.Context, pg *Postgres) (tx *sqlx.Tx, err error) {
+	// get or create tx
+	logger(ctx).Debug().
+		Msg("no transaction on context")
```

Comment:
this is duplicated before some of the calls to `createTx`

Reply:
resolved: https://github.com/brave-intl/bat-go/pull/793/commits/ca66e903529564cb9ad4e5af367a97fda7fa4f44